### PR TITLE
Add polars

### PR DESCRIPTION
**Pull request by Maxyme**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by dhirschfeld** [comment]
Super excited to see this! It seems like it's a very awkward one to compile for conda-forge though 😬.

If Windows is being particularly difficult perhaps this can be merged without Windows support and then debug it on the feedstock?

**Comment by Maxyme** [comment]
@dhirschfeld, just updated to skip windows builds for now, I think it will be quite simple to fix in a future PR. Thanks for the comment.

**Comment by dhirschfeld** [comment]
Seems like it can't find the python executable, which is odd since you explicitly specify `--interpreter=%PYTHON%` to the `maturin` command.
```
--- stderr 
error: failed to run the Python interpreter at C:\bld\polars_1652374306242\_build_env\python.exe: The system cannot find the file specified. (os error 2) 
warning: build failed, waiting for other jobs to finish... 
💥 maturin failed 
Caused by: Failed to build a native library through cargo 
Caused by: Cargo build finished with "exit code: 101": `cargo rustc --manifest-path Cargo.toml --message-format json --release --lib -- -C codegen-units=16 -C lto=thin -C target-cpu=native -C link-arg=-s` 
```

Maybe the selector means `python` isn't being installed in the build env? :thinking:
```yaml
      - python                              # [build_platform != target_platform]
```
Anyway, easy enough to test that theory out in the feedstock....

File: recipes/polars/build.sh
Comment (ritchie46) [inline recipes/polars/build.sh:55]: We should not be compiling for `native` target CPU. That would mean the binary is unusable for CPU's that don't share the same features.

For pypi we set: 

```
export RUSTFLAGS='-C target-feature=+fxsr,+sse,+sse2,+sse3,+ssse3,+sse4.1,+sse4.2,+popcnt,+avx,+fma'
```
This is a reasonable default.

Next, we also should not set the `codegen-units` nor `thin` lto. This creates a binary that would be slower than what we produce for pypi.

So I would propose skipping the whole `--rustc-extra-args="-C codegen-units=16 -C lto=thin -C target-cpu=native"` and setting:

```
export RUSTFLAGS='-C target-feature=+fxsr,+sse,+sse2,+sse3,+ssse3,+sse4.1,+sse4.2,+popcnt,+avx,+fma'
```

**Review by ritchie46 — COMMENTED** [review]
Awesome initiative! I've left a few comments regarding how we build the binary.

**Comment by timkpaine** [comment]
windows looks ok over here:https://github.com/Maxyme/staged-recipes/pull/1
will ping again once the builds are confirmed passing

File: recipes/polars/build.sh
Comment (timkpaine) [reply recipes/polars/build.sh]: we may have to modify this once the feedstock is created for mac arm builds

**Comment by sugatoray** [comment]
@BastianZim @beckermr @bollwyvl  Requesting your feedback on the recipe. 

The sources for `rust-nightly` are included in the recipe. Does it require the license(s) to be included for `rust-nightly` as well? Thank you.

**Review by xhochy — CHANGES_REQUESTED** [review]
Can we instead please update the rust_dev build of the rust conpiler feedstock instead of bundling it here?

**Comment by timkpaine** [comment]
@xhochy is there an existing pattern for using the rust_dev build like this? Just keeping in mind that this is the 5th attempt at getting a polars feedstock in and there's a dearth of prior rust-nightly projects

**Comment by xhochy** [comment]
> @xhochy is there an existing pattern for using the rust_dev build like this? 

Add the following to `conda_build_config.yaml`:

```
channel_sources:
 - conda-forge/label/rust_dev,conda-forge
```

You might need to update the rust_dev build in the rust-feedstock to get the latest version there.

> Just keeping in mind that this is the 5th attempt at getting a polars feedstock in and there's a dearth of rust-nightly examples to work from

The number of failed PRs should not be a wildcard to dismiss reviews. We want sustainable solutions here.

**Comment by timkpaine** [comment]
@xochy https://github.com/pola-rs/polars/issues/383#issuecomment-1124249872

> got most of my inspiration from ORJSON's feedstock (https://github.com/conda-forge/orjson-feedstock)

**Comment by xhochy** [comment]
The PR shouldn't have been merged. We package the compiler in https://github.com/conda-forge/rust-feedstock/blob/066e090a2b11c9fb43b14941d69c529b6d8b861e/recipe/meta.yaml#L1 That branch is outdated but updating it shouldn't be difficult.

**Comment by timkpaine** [comment]
@xhochy what about this? https://github.com/conda-forge/orjson-feedstock/issues/33#issuecomment-766534969

**Comment by timkpaine** [comment]
Polars is extremely fast moving, without automated rebuilding of rust-dev on every nightly release, I don't think it will be sustainable

**Comment by xhochy** [comment]
There is nothing stopping you from becoming a `rust` feedstock maintainer and updating the version there, too. We can move there as fast as in the `polars` feedstock itself.

**Comment by xhochy** [comment]
Otherwise, we will build the rust compiler in several other feedstocks in future, too. I would expect more packages to depend on `rust-nightly` in future.

**Comment by timkpaine** [comment]
Yes its already been a problem, and I think one of the reasons why its taken 5 tries for polars was because of the lack of standardization. I know in my PR I based it on a different (merged) feedstock that pulled nightly in a different manner

**Comment by timkpaine** [comment]
let me see how painful the process of a new rust-dev build is

**Comment by xhochy** [comment]
@Maxyme You will need `- {{ compiler('rust') }}` as a `build` requirement

**Comment by Maxyme** [comment]
@xhochy It looks like it still picks the 1.60 stable version if I add `- {{ compiler('rust') }}`. I'm not sure how to specify that we want the rust_dev version. I tried following this here: https://conda-forge.org/docs/maintainer/knowledge_base.html?highlight=channel_sources#creating-a-pre-release-build.

**Comment by xhochy** [comment]
The issue is that staged-recipes doesn't support `channel_sources` at the moment. I would suggest doing the following:

1. Temporarily add `conda-forge/labels/rust_dev` to https://github.com/conda-forge/staged-recipes/blob/a6a2c7c8410785c792ce55059ac6d15db5516dab/.ci_support/linux64.yaml#L10 and https://github.com/conda-forge/staged-recipes/blob/a6a2c7c8410785c792ce55059ac6d15db5516dab/.ci_support/osx64.yaml#L8
2. Get a green build to verify that it will work on the feedstock
3. Revert the commit made in 1, merge.

**Comment by Maxyme** [comment]
Getting this error now:
```
conda.exceptions.UnavailableInvalidChannel: The channel is not accessible or is invalid.
  channel name: conda-forge/labels/rust_dev
  channel url: https://conda.anaconda.org/conda-forge/labels/rust_dev
  error code: 404
```
from
`requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://conda.anaconda.org/conda-forge/labels/rust_dev/noarch/repodata.json`

**Comment by timkpaine** [comment]
Would it be crazy to publish rust-nightly directly on conda forge? Unlike other pre-release version, it seems many libraries treat rust-nightly as a production equivalent. And I expect many more python packages will have rust nightly dependencies, repeating this exercise isn't going to scale well.

File: .ci_support/linux64.yaml
Comment (jakirkham) [inline .ci_support/linux64.yaml:10]:
```suggestion
- conda-forge/label/rust_dev
```

File: .ci_support/osx64.yaml
Comment (jakirkham) [inline .ci_support/osx64.yaml:8]:
```suggestion
  - conda-forge/label/rust_dev
```

File: .ci_support/win64.yaml
Comment (jakirkham) [inline .ci_support/win64.yaml:2]:
```suggestion
  - conda-forge/label/rust_dev
```

**Review by jakirkham — COMMENTED** [review]
Should be `label` instead of `labels`

**Comment by Maxyme** [comment]
@timkpaine I agree about rust-nightly, it would be nice to have it as a separate compiler.

**Comment by jakirkham** [comment]
Agree with Isuru. That or ask Rust to make a release

**Comment by jakirkham** [comment]
Looks like one of the compilation errors was filed in upstream issue ( https://github.com/pola-rs/polars/issues/2894 )

The other compilation issues around simd am not seeing an issue for

**Comment by Maxyme** [comment]
@jakirkham compilation should work with rust-nightly-2022-04-14, which is what was merged recently in the rust_dev branch (same compiler I'm trying to use in the build env). I'll have to check what's happening.

**Comment by dhirschfeld** [comment]
Maybe this is one which should just be merged as-is and fixed in the actual feedstock?


Improving `staged-recipes` to handle different channels sounds like a great improvement but shouldn't necessarily be a blocker for getting this in IMHO.

File: recipes/polars/meta.yaml
Comment (jakirkham) [inline recipes/polars/meta.yaml:23]:
```suggestion
    - {{ compiler('rust') }}
    - conda-forge/label/rust_dev::rust
```

**Comment by Maxyme** [comment]
@dhirschfeld @xhochy In order to merge this, I can revert this PR to the commit which was installed rust-nightly directly. It may be easier to update/fix afterwards.

**Comment by xhochy** [comment]
No, I would like to see this being green here and merge then. @jakirkham's suggestion sounds good but it seems, it still picked up an old version. Retriggered it since we got some builds uploaded since the commit was pushed.

File: recipes/polars/meta.yaml
Comment (xhochy) [inline recipes/polars/meta.yaml:23]:
```suggestion
    - conda-forge/label/rust_dev::rust_{{ target_platform }}
```

File: .ci_support/osx64.yaml
Comment (xhochy) [inline .ci_support/osx64.yaml:9]:
```suggestion
  - conda-forge/label/rust_dev,conda-forge
```

File: .ci_support/win64.yaml
Comment (xhochy) [inline .ci_support/win64.yaml:3]:
```suggestion
  - conda-forge/label/rust_dev,conda-forge
```

File: recipes/polars/meta.yaml
Comment (xhochy) [inline recipes/polars/meta.yaml:24]:
```suggestion
    - conda-forge/label/rust_dev::rust >=1.62
```

File: recipes/polars/build.sh
Comment (isuruf) [inline recipes/polars/build.sh:5]: Why?

File: recipes/polars/bld.bat
Comment (isuruf) [inline recipes/polars/bld.bat:2]: Why `cpu=native`?

File: recipes/polars/build.sh
Comment (Maxyme) [reply recipes/polars/build.sh:5]: You can see the initial comment here: 
https://github.com/conda-forge/staged-recipes/pull/18906#pullrequestreview-971773615

File: recipes/polars/bld.bat
Comment (Maxyme) [reply recipes/polars/bld.bat:2]: I disabled windows builds for now, and I haven't updated the bld.bat to match the build.sh script.

File: recipes/polars/build.sh
Comment (isuruf) [reply recipes/polars/build.sh]: We don't enable `avx` by default on conda-forge binaries. So, please remove `RUSTFLAGS`.

File: recipes/polars/bld.bat
Comment (jakirkham) [reply recipes/polars/bld.bat:2]: Maybe we should drop this file then?

**Comment by isuruf** [comment]
See https://github.com/conda-forge/conda-forge.github.io/issues/1052 about licenses for rust projects.

**Comment by BastianZim** [comment]
> See [conda-forge/conda-forge.github.io#1052](https://github.com/conda-forge/conda-forge.github.io/issues/1052) about licenses for rust projects.

Summary is also now in the docs: https://conda-forge.org/docs/maintainer/adding_pkgs.html#licenses-of-included-dependencies

**Comment by Maxyme** [comment]
@isuruf @xhochy Thanks a lot for fixing the build! As I see it, there are 3 things left to do on this PR:

- [ ] Check if avx is needed as compilation flag (may be for SIMD instructions, I'm not sure)
- [ ] Remove bld.bat file
- [ ] Add cargo-bundle-licenses to bundle third-party licenses.

File: recipes/polars/bld.bat
Comment (timkpaine) [reply recipes/polars/bld.bat:2]: No just leave it, we plan to add windows shortly its just hitting some permissions weirdness and we don't want to block

Comment (xhochy) [reply recipes/polars/bld.bat:2]: Then you should use the same flags as on Linux and remove the `cpu-native`.

File: .ci_support/build_all.py
Comment (xhochy) [inline .ci_support/build_all.py:63]: Seems like the changes here are working, I would separate them out into a PR to staged-recipes tomorrow so we don't need to hack anything here.

File: recipes/polars/build.sh
Comment (isuruf) [inline recipes/polars/build.sh:6]: Remove this so that the default options are used and when conda-forge moves forward, it'll be done automatically.

File: recipes/polars/bld.bat
Comment (timkpaine) [reply recipes/polars/bld.bat]: will do

File: recipes/polars/build.sh
Comment (ritchie46) [reply recipes/polars/build.sh]: Note that we use these feature flags in polars. Or are they moved somewhere else? I am not really familiar with conda feedstocks.

Comment (ritchie46) [reply recipes/polars/build.sh]: Yes thats correct but these are reasonable defaults that practically all machines have. The rust defaults are super conservative because they also ship to embedded.

 Note that we don't activate avx. That one is not yet broadly supported.

Comment (Maxyme) [reply recipes/polars/build.sh]: Would you be okay going ahead with the defaults and looking at the flags in a separate pr? I can open a new one after this for that purpose. @isuruf ?

**Comment by timkpaine** [comment]
ive been duplicating our work in this PR on https://github.com/conda-forge/staged-recipes/pull/18987 so whenever we merge this one if we could do `connector-x` as well (same deal, rust nightly).

**Comment by Maxyme** [comment]
Small update: this PR is currently blocked by: https://github.com/conda-forge/staged-recipes/pull/19036

**Comment by jakirkham** [comment]
Now that PR ( https://github.com/conda-forge/staged-recipes/pull/19036 ) is in would it be possible to drop those changes from here?

**Comment by xhochy** [comment]
@Maxyme Can you rebase and remove all changes that aren't inside `recipe/polars`? Then we can finally merge :) 

Also I'm happy to be added as a maintainer to help with future build troubles if you would like to.

**Comment by dhirschfeld** [comment]
Thanks @xhochy for getting this over the line, and making it easier to add new rust recipes! :tada:

File: recipes/polars/meta.yaml
Comment (dhirschfeld) [inline recipes/polars/meta.yaml:63]: :point_up: I'm happy to be added as a maintainer

**Review by xhochy — APPROVED** [review]

**Comment by Maxyme** [comment]
@isuruf, ready to merge this!

**Comment by timkpaine** [comment]
(applying the same changes to https://github.com/conda-forge/staged-recipes/pull/18987)

**Comment by timkpaine** [comment]
moving to https://github.com/conda-forge/staged-recipes/pull/19136

**Review by dhirschfeld — APPROVED** [review]

**Comment by timkpaine** [comment]
🎉 🎉 🎉

**Pull request merged**
